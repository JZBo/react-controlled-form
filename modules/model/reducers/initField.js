/* @flow */
import type { Action } from '../../../types/Action'

import composeFieldData from '../../utils/composeFieldData'

export default function initField(
  state: Object,
  {
    payload: {
      formId,
      fieldId,
      value = '',
      isEnabled = true,
      isRequired = false,
      isValid = true
    }
  }: Action
): Object {
  return {
    ...state,
    [formId]: {
      ...state[formId],
      data: {
        ...state[formId].data,
        [fieldId]: composeFieldData(
          {
            value,
            isEnabled,
            isRequired,
            isTouched: false,
            isValid
          },
          state[formId].data[fieldId] || {}
        )
      }
    }
  }
}
