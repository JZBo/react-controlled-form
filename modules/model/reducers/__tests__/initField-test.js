import initField from '../initField'

const payload = p => ({ payload: p })
const initialFormState = { data: {}, state: {} }

describe('Initializing a form field', () => {
  it('should use default values', () => {
    expect(
      initField(
        { foo: initialFormState },
        payload({ formId: 'foo', fieldId: 'bar' })
      )
    ).toMatchSnapshot()
  })

  it('should overwrite default values', () => {
    expect(
      initField(
        { foo: initialFormState },
        payload({
          formId: 'foo',
          fieldId: 'bar',
          isValid: false,
          isRequired: true,
          value: true,
        })
      )
    ).toMatchSnapshot()
  })

  it('should keep initial form field values', () => {
    expect(
      initField(
        {
          foo: {
            data: {
              bar: {
                value: 'baz',
                isValid: false,
                isEnabled: false,
              },
            },
            state: {},
          },
        },
        payload({
          formId: 'foo',
          fieldId: 'bar',
          isValid: false,
          isRequired: true,
          value: true,
        })
      )
    ).toMatchSnapshot()
  })
})
