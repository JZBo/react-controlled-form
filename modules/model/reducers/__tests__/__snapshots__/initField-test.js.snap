// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Initializing a form field should keep initial form field values 1`] = `
Object {
  "foo": Object {
    "data": Object {
      "bar": Object {
        "_initial": Object {
          "isEnabled": false,
          "isRequired": true,
          "isTouched": false,
          "isValid": false,
          "value": "baz",
        },
        "isEnabled": false,
        "isRequired": true,
        "isTouched": false,
        "isValid": false,
        "value": "baz",
      },
    },
    "state": Object {},
  },
}
`;

exports[`Initializing a form field should overwrite default values 1`] = `
Object {
  "foo": Object {
    "data": Object {
      "bar": Object {
        "_initial": Object {
          "isEnabled": true,
          "isRequired": true,
          "isTouched": false,
          "isValid": false,
          "value": true,
        },
        "isEnabled": true,
        "isRequired": true,
        "isTouched": false,
        "isValid": false,
        "value": true,
      },
    },
    "state": Object {},
  },
}
`;

exports[`Initializing a form field should use default values 1`] = `
Object {
  "foo": Object {
    "data": Object {
      "bar": Object {
        "_initial": Object {
          "isEnabled": true,
          "isRequired": false,
          "isTouched": false,
          "isValid": true,
          "value": "",
        },
        "isEnabled": true,
        "isRequired": false,
        "isTouched": false,
        "isValid": true,
        "value": "",
      },
    },
    "state": Object {},
  },
}
`;
